### DISCUSSION TITLE
I’m unable to activate my oqs-provider

### DISCUSSION THREAD

**Discussion by XFhhPS** [open-quantum-safe/oqs-provider#391] [Q&A] [answered]
I’m unable to activate my oqs-provider, even though I’ve followed the instructions in the usage documentation to add the corresponding information to etc/ssl/openssl. I’ve also restarted the system.

I installed it on Ubuntu 20.04, and before installation, I backed up the system’s default openssl1.1.1f and built and installed openssl3.3.0-beta1. I created symbolic links and reloaded ldconfig, so the current system openssl version output is openssl3.3.0-beta1.

`xiaofan@xiaofan-virtual-machine:~/桌面/openssl-3.3.0-beta1$ openssl version 
OpenSSL 3.3.0-beta1 29 Mar 2024 (Library: OpenSSL 3.3.0-beta1 29 Mar 2024)
`

I built it manually, first building and installing liboqs, then using traditional cmake to build oqs-providers as unix.md. oqs-providers was successfully built, but it wasn’t installed correctly, so I manually moved oqs-providers.so to usr/local/openssl/lib64/module. This allows it to list post-quantum signatures and kem algorithms. The conmand can work.
`openssl list -signature-algorithms -provider oqsprovider`

However, I still can’t activate oqs-provider. I’ve added the corresponding information to etc/ssl/openssl.cnf as per the usage instructions, but it didn’t work.

> [provider_sect]
default = default_sect
oqsprovider = oqsprovider_sect
[oqsprovider_sect]
activate = 1

i have added but

> xiaofan@xiaofan-virtual-machine:/etc/ssl$ openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.3.0
    status: active

I also failed to load oqs-provider for pqc operations in a C program,
`
#include <openssl/provider.h>
#include <openssl/evp.h>
#include <openssl/err.h>
#include <stdio.h>

int main() {
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
    OSSL_PROVIDER *prov = OSSL_PROVIDER_load(libctx, "oqsprovider");
    EVP_PKEY *pkey = NULL;
`
encountering the following error.

> Failed to create EVP_PKEY_CTX
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80EBFB1D667F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (oqs_sig_default : 0), Properties (<null>)

What should I do?

**Comment by XFhhPS** [comment]
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/fc601e55-0937-4fe3-8f52-afa604b14077)
这是我的openssl.cnf文件

**Reply by baentsch** [reply]
If this command works OK listing all PQ algs: `openssl list -signature-algorithms -provider oqsprovider` and this command does not `openssl list -signature-algorithms`, then clearly the (apparently correct) "openssl.cnf" file isn't properly loaded. Please consider using OPENSSL_CONF env var to set that path explicitly or use a system tool like `lsof` to determine which config file your `openssl` executable uses.

**Reply by XFhhPS** [reply]
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/f70b326d-5435-40ac-a99a-1cd430729e78)
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/e96efa45-0633-434b-98b1-23fe82345774)
dear sir
at usr/local/openssl/bin only have the oqs-provider
at desktop only have the default-provider
why?

![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/bd3f9dac-94ac-4c38-b57b-3317aacb0d40)

I have set the OPENSSL_CONF env,but in desktop the command still not work. And in usr/local/openssl/bin `echo $OPENSSL_CONF ` 
doesn't return anything

**Reply by baentsch** [reply]
If 

> echo $OPENSSL_CONF 
doesn't return anything

the variable obviously has not been set properly. Then oqsprovider cannot be activated and the subsequent steps are a logical consequence. This has nothing to do with oqsprovider but with your local "desktop" setup I do not have access to and cannot comment on/debug.

**Reply by XFhhPS** [reply]
I think it’s not a logical consequence. 
I previously mentioned that only oqs-provider is displayed in usr/local/openssl/bin, and only oqs-provider is displayed in the ext/ssl path as well, 
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/989ef7d4-bc5d-424f-af2b-b3937b4073ee)

![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/871f4765-9137-4be6-9d4f-d8f4f8cae93b)
but only default-provider is displayed on the desktop. 
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/aa65c6af-85eb-450e-97ca-9550ce2e8735)

Moreover, it is normal for an environment variable exported in one terminal not to be displayed in another terminal. I used the command echo 

> 'export OPENSSL_CONF=/etc/ssl/openssl.cnf' >> ~/.bashrc 

to load the OPENSSL_CONF variable into /.bashrc, making it a global variable. Subsequently, oqs-provider no longer shows up in the ext/ssl path or in usr/local/openssl/bin, which I don’t think is a logical consequence.
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/608a53e2-ccd3-4fec-a09b-3777f0ced5d5)
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/bc52d927-2955-4bbf-aa3c-12bd988756e7)

**Reply by baentsch** [reply]
Well, in the above I assumed [OPENSSL_MODULES](https://github.com/open-quantum-safe/oqs-provider/blob/main/USAGE.md#module-option) has been configured correctly: Did you do that?

**Reply by XFhhPS** [reply]
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/d8bd3a51-b492-4cdb-ab9a-fb125b655929)
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/539c188f-e77e-4510-9389-0d6d5bc28087)
I have configured this env,but it does not make sense.

I also added module in openssl.cnf
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/edda1b3d-a3b6-443b-8600-03a5c0a68637)
but I still can't see oqsprovider activated

**Reply by XFhhPS** [reply]
`xiaofan@xiaofan-virtual-machine:~/桌面$ export OPENSSL_MODULES=/usr/local/openssl/lib64/ossl-modules
xiaofan@xiaofan-virtual-machine:~/桌面$ echo $OPENSSL_MODULES
/usr/local/openssl/lib64/ossl-modules
xiaofan@xiaofan-virtual-machine:~/桌面$ openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.3.0
    status: active

I changed the environment variable from the .so file to the directory, but it still doesn't work

**Reply by baentsch** [reply]
Please run 'ls $OPENSSL_MODULES' and 'cat $OPENSSL_CONF'

**Reply by XFhhPS** [reply]
Hello sir, I've managed to activate oqsprovider, but now the issue is that I cannot activate oqsprovider and default-provider simultaneously. When my OpenSSL configuration file is at etc/ssl/openssl.cnf, I cannot activate oqsprovider. When my OpenSSL configuration file is at usr/local/openssl/openssl.cnf, I can activate oqsprovider with the added configurations below:

![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/4dc975cf-59e1-4419-b533-d6f0dd761eeb)
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/62cf2a1c-c36e-4849-b0bc-f6193ea9b27a)

However, at this time, only oqsprovider is active without default-provider.

When I add:
`
[default_sect]
activate = 1
`
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/603ee62b-7e21-4111-a408-23196b61b071)

Afterwards, only default-provider is active without oqsprovider.
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/3f8ae133-91b4-4c93-8068-8f3411d81cb6)

I am really confused. What might be the reason for this?

**Reply by baentsch** [reply]
The syntax of your openssl.cnf file is wrong. Please compare with the [test one](https://github.com/open-quantum-safe/oqs-provider/blob/main/test/openssl-ca.cnf).

**Reply by XFhhPS** [reply, marked as answer]
It's amazing！I simply swapped the positions of [default_sect] and [oqsprovider_sect], and it worked! 
I changed it from 
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/4ef2a46a-4650-4069-9389-365c50dedef5)
 to 
 
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/09322d33-f1e1-4086-aa5b-102bebf2239f)

 and then it was fixed！
![image](https://github.com/open-quantum-safe/oqs-provider/assets/109072392/7cf11916-de47-4f57-b690-d1618af2faa0)
Thank you for your patience in answering my questions; I will continue to study the OQC project diligently.
